### PR TITLE
Don't add chaperone space after escaped quote in triple quote

### LINKED ISSUE #16640: Unexpected space in docstring near escaped quote
### Description

Ruff formatter modifies the spacing in a string within a docstring when the string contains an escaped quote. Specifically, it adds an unwanted space between the escaped quote and the closing triple quotes of the docstring.

### Steps to Reproduce

1. Create a Python class with the following content:
   ```python
   class Sample:
       """Hello "World\""""
   
       def __init__(self, name):
           self.name = name

2. Run `ruff format` on the file
3. Observe that the formatter modifies the docstring to:
    ```python
    class Sample:
        """Hello "World\" """
    
        def __init__(self, name):
            self.name = name

### Expected Behavior
Docstring remain unchanged after formatting as its escaped already

`"""Hello "World\""""`

### Version

ruff 0.9.10 (0dfa810e9 2025-03-07)

### PR DESCRIPTION
**Pull request by maxmynter**
Closes #16640

## Summary
Added a check for escaped quotes in triple quotes to determine if chaperone spaces are needed.
This prevents unwanted formatting of valid docstrings that end in escaped quotes.

## Test Plan
Added a unit test testing that docstrings ending in escaped double quotes are not reformatted.

### CODE REVIEW

File: crates/ruff_python_formatter/src/string/docstring.rs [hunk 1]
```diff
@@ -1597,11 +1597,14 @@ fn docstring_format_source(
...
 /// If the last line of the docstring is `content" """` or `content\ """`, we need a chaperone space
-/// that avoids `content""""` and `content\"""`. This does only applies to un-escaped backslashes,
-/// so `content\\ """` doesn't need a space while `content\\\ """` does.
+/// that avoids `content""""` and `content\"""`. This does only applies to un-escaped or escaping
+///  backslashes, so `content\\ """` or `content\""""` don't need a space while `content\\\ """` does.
 pub(super) fn needs_chaperone_space(flags: AnyStringFlags, trim_end: &str) -> bool {
-    if trim_end.chars().rev().take_while(|c| *c == '\\').count() % 2 == 1 {
-        true
+    if (flags.is_triple_quoted()
+        && trim_end.ends_with(&std::format!("\\{}", flags.quote_style().as_char())))
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/string/docstring.rs:1604]: I don't think adding the check to this branch makes sense because 

```
trim_end.chars().rev().take_while(|c| *c == '\\').count() % 2 == 1
```
will always return `false` if the string ends with `\'` or `\"` 

I think it also incorrectly returns `false` for 

```py
"""Hello "World\\""""
```

Comment (maxmynter) [reply crates/ruff_python_formatter/src/string/docstring.rs]: Thanks for flagging this; you're right. It seems i underestimated edge cases and then got caught up in premature refactoring without sufficient test cases.

I've added a bunch more tests in [d6e332b](https://github.com/astral-sh/ruff/pull/17216/commits/d6e332bf3a2fb6790f20e552e8db6feb89621ba8) and fixed the implementation in [5a5f55c](https://github.com/astral-sh/ruff/pull/17216/commits/5a5f55ce0cb7d383f35699243445f09a0ee12611).

File: crates/ruff_python_formatter/src/string/docstring.rs [hunk 2]
```diff
@@ -1596,15 +1596,27 @@ fn docstring_format_source(
...
+            return true;
+        }
+        if let Some(before_quote) = trim_end.get(..trim_end.len() - 1) {
+            if count_consecutive_chars_from_end(before_quote, '\\') % 2 == 0 {
+                // Even backslash count preceding quote; chaperone avoids dangling
+                return true;
+            }
+        }
     }
+    false
```
[the remark this answers is not in the document]
Comment (maxmynter) [reply crates/ruff_python_formatter/src/string/docstring.rs]: With slight adaptations, yes, we can simplify:
- used `strip_suffix` instead of `rstrip_prefix` (which doesn't exist).
- Check for even not odd backslash count before the closing quote.

Implemented in [ad8604f](https://github.com/astral-sh/ruff/pull/17216/commits/ad8604f6b09ef25916ef683881f0a51e20f1c097)

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_chaperones.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+class Sample:
+    """Needs chaperone\\" """
+
+    """Needs chaperone\\\ """
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_chaperones.py:4]: This file doesn't test the chaperon handling except for the first string because all subsequent strings aren't docstrings.
